File: framework/test2/common/intbreak/intbreak.cpp
#ifdef __clang__
    #include "soda/clang/stdc++.h"
#else
    #include <bits/stdc++.h>
#endif
#include "soda/leetcode/leetcode.h"
#include "soda/unittest/unittest.h"

using namespace std;
using namespace soda::leetcode;
using namespace soda::unittest;

// step [1]: implement class Solution
// class Solution {};
class Solution {
    vector<int> memo;
public:
    int integerBreak(int n) {
        memo.assign(59, 0);
        return solve(n);
    }

private:
    int solve(int n) {
        if (n == 1) {
            return 1;
        }
        if (memo[n] > 0) {
            return memo[n];
        }
        int res = 0;
        for (int i = 1; i < n; ++i) {
            res = std::max(i * (n-i), res);
            res = std::max(i * solve(n-i), res);
        }
        memo[n] = res;
        return res;
    }
};

namespace {
    const auto __ = []() {
        ios_base::sync_with_stdio(false);
        cin.tie(nullptr);
        return 0;
    }();
}

int main()
{
    // [1] create by class member function
    Solution su;
    auto work = WorkFactory::create(su, &Solution::integerBreak);
    //
    // [2] or, create by ordinary function
    // auto work = WorkFactory::create(function);

    // work->setValidator(validate);
    work->setCompareSerial(true);
    work->run();
    delete work;
    return 0;
}

File: framework/test2/common/intbreak/Intbreak.scala
import soda.scala.unittest._

import scala.reflect.runtime.universe._

class Solution {}
object Solution {
  var memo: Array[Int] = Array.empty
  def integerBreak(n: Int): Int = {
    memo = Array.fill(59)(0)
    solve(n)
  }

  def solve(n: Int): Int = {
    if (n == 1) return 1
    if (memo(n) > 0) return memo(n)
    var res = 0
    for (i <- 1 until n) {
      res = res max (i * (n-i))
      res = res max (i * solve(n-i))
    }
    memo(n) = res
    res
  }
}

class Intbreak {
  def get(): ()=>Unit = {
    // val work = TestWork.forObject(typeOf[Solution], "integerBreak")
    // val work = GenericTestWork.create1(Solution.integerBreak)
    // val work = TestWork.create1(Solution.integerBreak)
    () => {
      var work = GenericTestWork.create1(Solution.integerBreak)
      work.compareSerial = true
      work.run()
    }
  }
}

object Intbreak extends App {
  new Intbreak().get().apply()
}
